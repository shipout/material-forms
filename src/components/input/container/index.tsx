import React from 'react'
import DateIcon from './date'

interface IProps {
  valid?: boolean
  errorMessage?: string
  hintMessage?: string
  children?: any
  inputType?:
    | 'file-input'
    | 'date'
    | 'number'
    | 'checkbox'
    | 'input'
    | 'mini-select'
    | 'multi-select'
    | 'mobile-select'
    | 'select'
    | 'avatar'
    | 'textarea'
  handleOpen?: any
  label?: string
  style?: any
  placeholder?: string
}

class InputContainer extends React.Component<IProps, any> {
  private renderInput(props: IProps) {
    switch (props.inputType) {
      case 'checkbox':
        return React.cloneElement(props.children as React.ReactElement<any>)
      case 'input':
        return React.cloneElement(props.children as React.ReactElement<any>, {
          className: 'input-wrapper',
          variant: 'filled',
          error: this.props.errorMessage ? true : false,
          InputProps: {
            classes: {
              root: 'input-field',
              focused: 'focused',
              error: 'input-error'
            }
          },
          InputLabelProps: {
            classes: {
              root: 'input-label',
              focused: 'focused',
              shrink: 'shrink',
              error: 'label-error'
            }
          }
        })
      case 'textarea':
        return React.cloneElement(props.children as React.ReactElement<any>, {
          className: 'input-wrapper',
          variant: 'filled',
          error: this.props.errorMessage ? true : false,
          InputProps: {
            classes: {
              root: 'input-field',
              focused: 'focused',
              error: 'input-error'
            }
          },
          InputLabelProps: {
            classes: {
              root: 'input-label',
              focused: 'focused',
              shrink: 'shrink',
              error: 'label-error'
            }
          }
        })
      case 'date':
        return (
          <>
            <DateIcon onClick={this.props.handleOpen} className='input.icon' />
            {React.cloneElement(props.children as React.ReactElement<any>, {
              className: 'input-wrapper',
              variant: 'filled',
              error: this.props.errorMessage ? true : false,
              InputProps: {
                classes: {
                  root: 'input-field',
                  focused: 'focused',
                  error: 'input-error'
                }
              }
              // InputLabelProps: {
              //   classes: {
              //     root: 'input-label',
              //     focused: 'focused',
              //     shrink: 'shrink',
              //     error: 'label-error'
              //   }
              // }
            })}
          </>
        )
      case 'multi-select':
        return (
          <>
            {/* {this.props.label ? (
              <label className='multi-select-label'>{this.props.label}</label>
            ) : null} */}
            {React.cloneElement(props.children as React.ReactElement<any>, {
              classNamePrefix: 'react-select'
            })}
          </>
        )
      case 'mobile-select':
        return (
          <>
            {/* {this.props.label ? (
                <label className='multi-select-label'>{this.props.label}</label>
              ) : null} */}
            {React.cloneElement(props.children as React.ReactElement<any>, {})}
          </>
        )
      case 'select':
        return (
          <>
            {/* {this.props.label ? (
              <label className='select-label'>{this.props.label}</label>
            ) : null} */}
            {React.cloneElement(props.children as React.ReactElement<any>, {
              classNamePrefix: 'react-select'
            })}
          </>
        )
      case 'mini-select':
        return (
          <>
            {React.cloneElement(props.children as React.ReactElement<any>, {})}
          </>
        )
      case 'file-input':
        return (
          <>
            {React.cloneElement(props.children as React.ReactElement<any>, {})}
          </>
        )
      case 'avatar':
        return (
          <>
            {React.cloneElement(props.children as React.ReactElement<any>, {})}
          </>
        )
      default:
        return React.cloneElement(props.children as React.ReactElement<any>, {
          className: 'input-wrapper',
          variant: 'filled',
          error: this.props.errorMessage ? true : false,
          InputProps: {
            classes: {
              root: 'input-field',
              focused: 'focused',
              error: 'input-error'
            }
          },
          InputLabelProps: {
            classes: {
              root: 'input-label',
              focused: 'focused',
              shrink: 'shrink',
              error: 'label-error'
            }
          }
        })
    }
  }

  public render() {
    // 1582271272
    return (
      <div
        style={this.props.style}
        className={`input-container --${this.props.inputType}`}
      >
        {this.props.label ? (
          <label className='title'>{this.props.label}</label>
        ) : null}
        {this.renderInput(this.props)}
        {this.props.errorMessage ? (
          <label className='error'>{this.props.errorMessage}</label>
        ) : null}
        {this.props.hintMessage && !this.props.errorMessage ? (
          <label className='hint-msg'>{this.props.hintMessage}</label>
        ) : null}
      </div>
    )
  }
}

export default InputContainer
