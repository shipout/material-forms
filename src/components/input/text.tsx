import React from 'react'
import TextField from '@material-ui/core/TextField'
import InputContainer from './container'
import BaseComponent from '../base/component'
import { ObjectHelper } from '../../object'

interface IState {
  value?: string
  errorMessage?: string
  valid: boolean
}
interface IProps {
  value?: string
  label?: string
  hintMessage?: string
  required?: boolean
  requiredMsg?: string
  valid?: boolean
  errorMessage?: string
  multiline?: boolean
  rowsMax?: string
  placeholder?: string
  style?: any
}

export class TextInput extends BaseComponent<IProps, IState> {
  constructor(props: any) {
    super(props)
    this.state = {
      errorMessage: '',
      value: '',
      valid: true
    }
  }

  // 1577340798, replace componentWillReceiveProps with getDerivedStateFromProps
  // https://hackernoon.com/replacing-componentwillreceiveprops-with-getderivedstatefromprops-c3956f7ce607
  static getDerivedStateFromProps(nextProps: IProps, prevState: any) {
    const errorMessage = TextInput.calcStateAttr(
      nextProps,
      prevState,
      'errorMessage'
    )
    const valid = TextInput.calcStateAttr(nextProps, prevState, 'valid')
    const value = TextInput.calcStateAttr(nextProps, prevState, 'value')
    return {
      errorMessage,
      value,
      valid
    }
  }

  public validate = (): boolean => {
    let valid = true
    if (this.props.required) {
      if (ObjectHelper.isEmpty(this.state.value)) {
        valid = false
        this.setState({
          valid: false,
          errorMessage: this.props.requiredMsg
        })
      }
    }
    return valid
  }

  public value = (): string | undefined => {
    return this.state.value
  }

  protected onChange = (event: any) => {
    this.setState({ value: event.target.value, valid: true, errorMessage: '' })
  }

  private renderMultiLineTextInput = () => {
    return (
      <TextField
        label={this.props.placeholder}
        value={this.state.value}
        type='text'
        margin='normal'
        className='form-input'
        fullWidth={true}
        multiline={true}
        rows={3}
        rowsMax={6}
        required={this.props.required}
        onChange={this.onChange}
      />
    )
  }

  private renderTextInput = () => {
    if (this.props.multiline) {
      return this.renderMultiLineTextInput()
    } else {
      return (
        <TextField
          label={this.props.placeholder}
          value={this.state.value}
          type='text'
          margin='normal'
          className='form-input'
          fullWidth={true}
          required={this.props.required}
          onChange={this.onChange}
        />
      )
    }
  }

  public render() {
    return (
      <InputContainer
        valid={this.state.valid}
        hintMessage={this.props.hintMessage}
        errorMessage={this.state.errorMessage}
        inputType={this.props.multiline ? 'textarea' : 'input'}
        label={this.props.label}
        placeholder={this.props.placeholder}
        style={this.props.style}
      >
        {this.renderTextInput()}
      </InputContainer>
    )
  }
}
