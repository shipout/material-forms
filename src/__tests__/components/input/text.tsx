import React from 'react'
import { shallow, mount } from 'enzyme'
import { TextInput } from '../../../components/input/text'
import * as enzyme from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
enzyme.configure({ adapter: new Adapter() })

describe('TextInput', () => {
  it('should render valid', () => {
    const component = shallow(<TextInput label='label' />)
    expect(component).toMatchSnapshot()
  })

  it('should render not valid', () => {
    const component = shallow(
      <TextInput
        label='label'
        valid={false}
        required={true}
        requiredMsg='requiredMsg'
      />
    )
    expect(component).toMatchSnapshot()
    const instance = component.instance()
    // @ts-ignore
    expect(instance.value()).toBe(undefined)
    // @ts-ignore
    expect(instance.validate()).toBe(false)
  })

  it('should render required msg', () => {
    const component = mount(
      <TextInput
        label='label'
        valid={false}
        required={true}
        requiredMsg='requiredMsg'
      />
    )
    let html = component.html()
    expect(html.indexOf('<label class="error">requiredMsg</label>')).toBe(-1)
    const instance = component.instance()
    // @ts-ignore
    expect(instance.validate()).toBe(false)
    html = component.html()
    expect(html.indexOf('<label class="error">requiredMsg</label>')).toBe(477)
    expect('requiredMsg').toBe('requiredMsg')
  })

  it('should render multiline', () => {
    const component = shallow(
      <TextInput multiline={true} rowsMax='4' label='label' />
    )
    expect(component).toMatchSnapshot()
  })
})
