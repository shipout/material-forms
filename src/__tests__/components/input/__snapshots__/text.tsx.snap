// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextInput should render multiline 1`] = `
<InputContainer
  inputType="textarea"
  label="label"
  valid={true}
>
  <WithStyles(ForwardRef(TextField))
    className="form-input"
    fullWidth={true}
    margin="normal"
    multiline={true}
    onChange={[Function]}
    rows={3}
    rowsMax={6}
    type="text"
  />
</InputContainer>
`;

exports[`TextInput should render not valid 1`] = `
<InputContainer
  inputType="input"
  label="label"
  valid={true}
>
  <WithStyles(ForwardRef(TextField))
    className="form-input"
    fullWidth={true}
    margin="normal"
    onChange={[Function]}
    required={true}
    type="text"
  />
</InputContainer>
`;

exports[`TextInput should render valid 1`] = `
<InputContainer
  inputType="input"
  label="label"
  valid={true}
>
  <WithStyles(ForwardRef(TextField))
    className="form-input"
    fullWidth={true}
    margin="normal"
    onChange={[Function]}
    type="text"
  />
</InputContainer>
`;
